lib/ide/ide/src/controller.rs

```diff
@@ -0,0 +1,377 @@
+    }
+
+    impl SetupConfiguration {
+        /// Establishes connection with the remote file manager endpoint.
+        pub async fn connect_fm(&self) -> MockTransport {
+            // TODO [mwu] should not return mock transport but a real class
+            //      implementing the websocket-based transport.
+            todo!()
+        }
+    }
+}
+
+mod project {
+    //! Project controller.
+    use super::*;
+    use json_rpc::Transport;
+
+    /// Create a new project controller.
+    pub fn new(file_manager_transport:impl Transport + 'static) -> StrongHandle<project::Data> {
+        let file_manager = file_manager_client::Client::new(file_manager_transport);
+        let ret = Data {
+            file_manager,
+            modules: default(),
+        };
+        strong(ret)
+    }
+
+    /// Project controller's state.
+    #[derive(Debug)]
+    pub struct Data {
+        file_manager : file_manager_client::Client,
+        modules      : HashMap<module::Location, module::WeakHandle>,
+    }
+
+    impl Data {
+        /// Returns handle to given module's controller if already present.
+        pub fn lookup_module(&mut self, loc:&module::Location) -> Option<module::StrongHandle> {
+            let weak = self.modules.get(loc)?;
+            match weak.upgrade() {
+                Some(strong) => Some(strong),
+                None => {
+                    self.modules.remove(loc);
+                    None
+                }
+            }
+        }
+        /// Stores given module controller handle and returns it.
+        ///
+        /// Note: handle stored in the project controller is weak, so the caller
+        /// remains responsible for managing the module's controller lifetime.
+        pub fn insert_module(&mut self, data:module::Data) -> module::StrongHandle {
+            let path   = data.path.clone();
+            let module = strong(data);
+            // TODO what if already exists?
+            self.modules.insert(path, module::WeakHandle{ data: Rc::downgrade(&module)});
+            module
+        }
+
+        /// Processes event from file manager.
+        pub fn process_file_event(&mut self, _event:file_manager_client::Event) {
+            // TODO push event to the appropriate module's input notification channel
+            todo!()
+        }
+
+        /// Identifies module affected by given file manager's event.
+        pub fn relevant_module(event:&file_manager_client::Event) -> Option<module::Location> {
+            use file_manager_client::Event;
+            use file_manager_client::Notification::FilesystemEvent;
+            match event {
+                Event::Closed          => None,
+                Event::Error(_)        => None,
+                Event::Notification(n) => match n {
+                    FilesystemEvent(e) => Some(module::Location(e.path.0.clone()))
+                }
+            }
+        }
+    }
+
+    make_weak_handle!(WeakHandle,Data);
+
+    impl WeakHandle {
```

Comment:
I think operating on Strong handles is more common, so I would implement all the logic on strong handles and having the weak handle we can just upgrade it.